src/clib/pio_msg.c

```diff
@@ -0,0 +1,2179 @@
+    if ((ret = PIOc_inq_att(ncid, varid, name5, xtypep, lenp)))
+	return ret;
+
+    return PIO_NOERR;
+}
+
+/** Handle attribute inquiry operations. This code only runs on IO
+ * tasks.
+ *
+ * @param ios pointer to the iosystem_desc_t.
+ * @param msg the message sent my the comp root task.
+ * @return PIO_NOERR for success, error code otherwise.
+*/
+int inq_attname_handler(iosystem_desc_t *ios)
+{
+    int ncid;
+    int varid;
+    int attnum;
+    char name[NC_MAX_NAME + 1], *namep = NULL;
+    char name_present;
+    int mpierr;
+    int ret;
+
+    LOG((1, "inq_att_name_handler"));
+
+    /* Get the parameters for this function that the the comp master
+     * task is broadcasting. */
+    if ((mpierr = MPI_Bcast(&ncid, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&varid, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&attnum, 1, MPI_INT,  ios->compmaster, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&name_present, 1, MPI_CHAR, 0, ios->intercomm)))
+	return PIO_EIO;
+    LOG((2, "inq_attname_handler got ncid = %d varid = %d attnum = %d name_present = %d",
+	 ncid, varid, attnum, name_present));
+
+    /* Match NULLs in collective function call. */
+    if (name_present)
+	namep = name;
+
+    /* Call the function to learn about the attribute. */
+    if ((ret = PIOc_inq_attname(ncid, varid, attnum, namep)))
+	return ret;
+
+    return PIO_NOERR;
+}
+
+/** Handle attribute inquiry operations. This code only runs on IO
+ * tasks.
+ *
+ * @param ios pointer to the iosystem_desc_t.
+ * @param msg the message sent my the comp root task.
+ * @return PIO_NOERR for success, error code otherwise.
+*/
+int inq_attid_handler(iosystem_desc_t *ios)
+{
+    int ncid;
+    int varid;
+    int attnum;
+    char *name;
+    int namelen;
+    int id, *idp = NULL;
+    char id_present;
+    int mpierr;
+    int ret;
+
+    LOG((1, "inq_attid_handler"));
+
+    /* Get the parameters for this function that the the comp master
+     * task is broadcasting. */
+    if ((mpierr = MPI_Bcast(&ncid, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&varid, 1, MPI_INT, 0, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&namelen, 1, MPI_INT,  ios->compmaster, ios->intercomm)))
+	return PIO_EIO;
+    if (!(name = malloc((namelen + 1) * sizeof(char))))
+	return PIO_ENOMEM;
+    if ((mpierr = MPI_Bcast(name, namelen + 1, MPI_CHAR,  ios->compmaster, ios->intercomm)))
+	return PIO_EIO;
+    if ((mpierr = MPI_Bcast(&id_present, 1, MPI_CHAR, 0, ios->intercomm)))
+	return PIO_EIO;
+    LOG((2, "inq_attid_handler got ncid = %d varid = %d attnum = %d id_present = %d",
+	 ncid, varid, attnum, id_present));
+
+    /* Match NULLs in collective function call. */
+    if (id_present)
+	idp = &id;
+
+    /* Call the function to learn about the attribute. */
+    if ((ret = PIOc_inq_attid(ncid, varid, name, idp)))
```

Comment:
The id is bcast at the end of PIOc_int_attid. None of the msg functions have to bcast their results, that is always done by the PIOc_ functions.